core/src/main/java/io/aiven/kafka/tieredstorage/fetch/index/MemorySegmentIndexesCache.java

```diff
@@ -0,0 +1,131 @@
+
+    // for testing
+    RemovalListener<SegmentIndexKey, byte[]> removalListener() {
+        return (key, content, cause) -> log.debug("Deleted cached value for key {} from cache."
+            + " The reason of the deletion is {}", key, cause);
+    }
+
+    private static Weigher<SegmentIndexKey, byte[]> weigher() {
+        return (key, value) -> value.length;
+    }
+
+    protected AsyncCache<SegmentIndexKey, byte[]> buildCache(final CacheConfig config) {
+        final Caffeine<Object, Object> cacheBuilder = Caffeine.newBuilder();
+        config.cacheSize().ifPresent(maximumWeight -> cacheBuilder.maximumWeight(maximumWeight).weigher(weigher()));
+        config.cacheRetention().ifPresent(cacheBuilder::expireAfterAccess);
+        final var cache = cacheBuilder.evictionListener(removalListener())
+            .scheduler(Scheduler.systemScheduler())
+            .executor(executor)
+            .recordStats(() -> statsCounter)
+            .buildAsync();
+        statsCounter.registerSizeMetric(cache.synchronous()::estimatedSize);
+        return cache;
+    }
+
+    @Override
+    public InputStream get(
+        final ObjectKey objectKey,
+        final IndexType indexType,
+        final Supplier<byte[]> indexSupplier
+    ) throws StorageBackendException, IOException {
+        try {
+            return cache.asMap()
+                .compute(new SegmentIndexKey(objectKey, indexType), (key, val) -> {
+                    if (val == null) {
+                        statsCounter.recordMiss();
+                        return CompletableFuture.supplyAsync(indexSupplier, executor);
+                    } else {
+                        statsCounter.recordHit();
+                        return val;
+                    }
+                })
+                .thenApplyAsync(ByteArrayInputStream::new, executor)
+                .get(GET_TIMEOUT_SEC, TimeUnit.SECONDS);
+        } catch (final ExecutionException e) {
+            // Unwrap previously wrapped exceptions if possible.
+            Throwable cause = e.getCause();
+            if (cause instanceof RuntimeException) {
+                cause = cause.getCause();
+            }
+
+            // We don't really expect this case, but handle it nevertheless.
+            if (cause == null) {
+                throw new RuntimeException(e);
+            }
+
+            if (cause instanceof StorageBackendException) {
+                throw (StorageBackendException) cause;
+            }
+            if (cause instanceof IOException) {
+                throw (IOException) cause;
+            }
+
+            throw new RuntimeException(e);
+        } catch (final InterruptedException | TimeoutException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void configure(final Map<String, ?> configs) {
+        final var config = new CacheConfig(new ConfigDef(), configs, DEFAULT_MAX_SIZE_BYTES);
```

Comment:
Do we really need to instantiate `ConfigDef` here? Maybe move it inside `CacheConfig`?